### ISSUE TITLE
New rule proposal: Rewrite single-line functions as expression-bodied function members

### ISSUE THREAD

**Issue by nbarbettini** [DotNetAnalyzers/StyleCopAnalyzers#1525]
**Category**: Readability
**Rule**: Functions where the only line starts with `return` can be rewritten as C# 6.0 expression-bodied function members
Rationale: Improve readability and maintainability of the code

Bad:

```csharp
public string MyFavoriteNumber(int number)
{
    return $"My favorite number is {number}!";
}
```

Good:

```csharp
public string MyFavoriteNumber(int number) => $"My favorite number is {number}!";
```

Interested in hearing discussion on this one. I like the expression-bodied style a lot myself, but I'm not sure everyone feels the same way. Thoughts?

**Comment by sharwell** [comment]
I find expression-bodied members great for some things, such as [these simple members](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/Settings/ObjectModel/DocumentationSettings.cs#L137-L150), but much worse (IMO) than traditional methods for others [like these](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs#L57-L59).

Right now I'm :-1: on this rule for the following reasons:

1. There is no obvious complexity under which I can say expression-bodied members are better and over which I can say full members are better.
2. I'm still uncomfortable adding rules requiring the use of C# 6 features. I'm not sure these features have existed long enough in practice to declare them the de-facto standard for some implementation detail. In addition, rules which require C# 6 features may limit the ability of some users to adopt StyleCop's rules.

**Comment by vweijsters** [comment]
:-1: For me there is too much uncertainty if this would cover all described cases with an equal level of success. I would prefer the full members in most cases, because those are easier to modify when an additional statement is necessary to solve a problem.

**Comment by nbarbettini** [comment]
Cheers! Thanks for the feedback guys. Good points.

**Comment by GregReddick** [comment]
I certainly would not like this being on by default. I think the "bad" syntax is easier to read and maintain. Is there a good reference for why they supported this in c#6 in the first place?
